resource-management/pkg/aggregrator/list-watch-impl.go

```diff
@@ -0,0 +1,153 @@
+			var regionNodeEvents [][]*event.NodeEvent
+			var length uint64
+			var eventProcess bool
+
+			// create client to resource region manager
+			c := NewSimClient(Config{ServiceUrl: a.urls[i], RequestTimeout: 30 * time.Minute})
+
+			klog.V(3).Infof("Starting loop list-watching nodes from region: %v", a.urls[i])
+
+			// hack, list all 1m node in one call without pagination
+			regionNodeEvents, crv, length = a.listNodes(c, ListOptions{Limit: 1000000})
+
+			if length != 0 {
+				klog.V(4).Infof("Total (%v) region node events are listed successfully in (%v) RPs", length, len(regionNodeEvents))
+			} else {
+				// TODO: handel empty list
+			}
+
+			// Convert 2D array to 1D array
+			minRecordNodeEvents := make([]*event.NodeEvent, 0, length)
+			for j := 0; j < len(regionNodeEvents); j++ {
+				minRecordNodeEvents = append(minRecordNodeEvents, regionNodeEvents[j]...)
+			}
+
+			start := time.Now()
+			eventProcess, crv = a.EventProcessor.ProcessEvents(minRecordNodeEvents)
+			end := time.Now()
+			klog.V(6).Infof("Event Processor Processed nodes results : %v. duration: %v", eventProcess, end.Sub(start))
+
+			// start watch node changes
+			a.watchNodes(c, crv)
+
+		}(i)
+	}
+
+	klog.V(3).Infof("Finished for loop to connect to to resource region manager...")
+	return nil
+}
+
+func (a *Aggregator) listNodes(client SimInterface, listOpts ListOptions) ([][]*event.NodeEvent, types.TransitResourceVersionMap, uint64) {
+	var start, end time.Time
+
+	klog.Infof("List resources from region manager ...")
+	start = time.Now().UTC()
+	nodeList, crv, length, err := client.List(listOpts)
+	end = time.Now().UTC()
+	if err != nil {
+		klog.Errorf("failed list resource from region manager. error %v", err)
+		return nil, crv, 0
+	}
+	klog.V(3).Infof("Got [%v] RPs, [%v] nodes from region manager, list duration: %v", len(nodeList), length, end.Sub(start))
+
+	if metrics.ResourceManagementMeasurement_Enabled {
+		for i := 0; i < len(nodeList); i++ {
+			for j := 0; j < len(nodeList[i]); j++ {
+				if nodeList[i][j] != nil {
+					nodeList[i][j].SetCheckpoint(metrics.Aggregator_Received)
+				}
+			}
+		}
+	}
+
+	return nodeList, crv, length
+}
+
+func (a *Aggregator) watchNodes(client SimInterface, crv types.TransitResourceVersionMap) {
+	var start, end time.Time
+
+	klog.Infof("Watch resources update from region manager ...")
+	start = time.Now().UTC()
+	watcher, err := client.Watch(crv)
+	if err != nil {
+		klog.Errorf("failed list resource from region manager. error %v", err)
+	}
+
+	watchCh := watcher.ResultChan()
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer utilruntime.HandleCrash()
+		// retrieve updates from watcher
+		for {
+			select {
+			case record, ok := <-watchCh:
+				if !ok {
+					// End of results.
+					klog.Infof("End of results")
+					return
+				}
+
+				klog.V(9).Infof("Got node event from region manager, nodeId: %v", record.Node.Id)
+				a.processNode(&record)
```

Comment:
ListNodes has no processNode(&record),  only watchNodes has processNode(&record), right?

Reply:
yes, the processEvnents for listed nodes are in the Run() function. ie. code flow as:

listnodes
processEvents for listed nodes and get CRV
watch node changes with the CRV
processEvents for watched nodes